### PR TITLE
BLADE-316 Add new function and command for blade server

### PR DESCRIPTION
**Pull request by gamerson**

### CODE REVIEW

**Comment by christopherbryanboyd** [comment]
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import java.util.jar.Attributes;
+import java.util.jar.Manifest;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipFile;
+
+/**
+ * @author Simon Jiang
+ */
+public class JavaUtil {
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java]: What about `JarUtil` instead?

File: cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+				return null;
+			}
+		}
+
+		return null;
+	}
+
+	public static String getManifestProperty(File manifestFile, String propertyName) {
+		try {
+			String contents = BladeUtil.readStreamToString(Files.newInputStream(manifestFile.toPath()));
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java]: Instead of reading the contents of the `File` reference as a `String` and then converting into `ByteArrayInputStream`, can we do something like this:
```
try (JarInputStream inputStream = new JarInputStream(new FileInputStream(manifestFile))) {

Manifest mf = inputStream.getManifest();
```

File: cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java [hunk 3]
```diff
@@ -33,4 +51,56 @@
...
 
+	public static boolean isEmpty(Collection<?> collection) {
+		if ((collection == null) || collection.isEmpty()) {
+			return true;
+		}
+
+		return false;
+	}
+
+	public static boolean isEmpty(List<?> list) {
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java]: `List` implements `Collection`, so we should not need this method.

File: cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java [hunk 4]
```diff
@@ -33,4 +51,56 @@
...
+
+	public static boolean isEmpty(Object[] array) {
+		if ((array == null) || (array.length == 0)) {
+			return true;
+		}
+
+		return false;
+	}
+
+	public static boolean isEmpty(Set<?> set) {
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java]: `Set` implements `Collection`, so we should not need this method.

File: cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java [hunk 5]
```diff
@@ -33,4 +51,56 @@
...
+		}
+
+		return false;
+	}
+
+	public static boolean isNotEmpty(Collection<?> collection) {
+		return !isEmpty(collection);
+	}
+
+	public static boolean isNotEmpty(List<?> list) {
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java]: We should not need this method when we have `isNotEmpty(Collection<?> collection)`

File: cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java [hunk 6]
```diff
@@ -33,4 +51,56 @@
...
+		return !isEmpty(list);
+	}
+
+	public static boolean isNotEmpty(Object[] array) {
+		return !isEmpty(array);
+	}
+
+	public static boolean isNotEmpty(Set<?> set) {
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/cli/util/ListUtil.java]: We should not need this method when we have `isNotEmpty(Collection<?> collection)`

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+/**
+ * @author Simon Jiang
+ */
+public abstract class AbstractPortalBundleFactory implements PortalBundleFactory {
+
+	@Override
+	public Path canCreateFromPath(Map<String, Object> appServerProperties) {
+		Path retval = null;
+
+		String appServerPath = (String)appServerProperties.get("app.server.dir");
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: This might be safer
```
String appServerPath = String.valueOf(appServerProperties.get("app.server.dir"));
```

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+			retval = appServerLocation;
+		}
+
+		Path liferayHome = _detectLiferayHome(liferayHomelocation, true);
+
+		if (liferayHome != null) {
+			File[] directories = FileUtil.getDirectories(liferayHome.toFile());
+
+			for (File directory : directories) {
+				Path dirPath = Paths.get(directory.getPath());
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: This would be simpler:
```
Path dirPath = directory.toPath();
```

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+	public Path canCreateFromPath(Path location) {
+		Path retval = null;
+
+		Path liferayHome = _detectLiferayHome(location, true);
+
+		if (liferayHome != null) {
+			File[] directories = FileUtil.getDirectories(liferayHome.toFile());
+
+			for (File directory : directories) {
+				Path dirPath = Paths.get(directory.getPath());
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: This would be simpler:
```
Path dirPath = directory.toPath();
```

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+
+	private Path _detectLiferayHome(Path path, boolean checkParents) {
+		if (FileUtil.notExists(path)) {
+			return null;
+		}
+
+		Path osgiPath = Paths.get(path.toString(), "osgi");
+		Path licensePath = Paths.get(path.toString(), "license");
+		Path dataPath = Paths.get(path.toString(), "data");
+
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: How about 
```
Path osgiPath = path.resolve("osgi");
Path licensePath = path.resolve("license");
Path dataPath = path.resolve("data");
```

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+			}
+		}
+	}
+
+	public static List<String> readRefs(File file) throws IOException {
+		if (FileUtil.notExists(file)) {
+			return Collections.emptyList();
+		}
+
+		InputStream is = Files.newInputStream(file.toPath());
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: Can we wrap this instead as
```
try (InputStream is = Files.newInputStream(file.toPath()) {
	return readRefs(is);
}
```
That way we don't need the finally block

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+		File layersList = new File(repoRoot, "layers.conf");
+
+		if (FileUtil.notExists(layersList)) {
+			return new LayersConfig();
+		}
+
+		Reader reader = null;
+
+		try (InputStream inputStream = Files.newInputStream(layersList.toPath())) {
+			reader = new InputStreamReader(inputStream, "UTF-8");
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: The `InputStreamReader` can be a part of the `try-with-resources` block along with the `InputStream`, so the finally block would not be necessary.

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+		}
+
+		private LayersConfig(Properties properties) {
+			_configured = true;
+			_layersPath = _DEFAULT_LAYERS_PATH;
+			_addOnsPath = _DEFAULT_ADD_ONS_PATH;
+
+			boolean excludeBase = false;
+
+			String layersProp = (String)properties.get("layers");
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: This does not need to be cast to a `String`

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+				if (excludeBase) {
+					_layers = Collections.emptyList();
+				}
+				else {
+					_layers = Collections.singletonList("base");
+				}
+			}
+			else {
+				String[] layerNames = layersProp.split(",");
+				_layers = new ArrayList<>();
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: Could we shorten this to:
```
		_layers =  new ArrayList<>(Arrays.asList(layerNames));
		boolean hasBase = _layers.contains("base");
```
in order to avoid the for loop?

File: cli/src/main/java/com/liferay/blade/server/PortalJBossBundle.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+		return logPath.resolve("server.log");
+	}
+
+	@Override
+	public Path getAppServerPortalDir() {
+		Path retval = null;
+
+		if (bundlePath != null) {
+			retval = bundlePath.resolve("/standalone/deployments/ROOT.war");
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/PortalJBossBundle.java]: This might be more cross-platform as:
```
retval = bundlePath.resolve(Paths.get("standalone", "deployments", "ROOT.war"));
```

File: cli/src/main/java/com/liferay/blade/server/PortalWildFlyBundleFactory.java [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+		slot = slot == null ? "main" : slot;
+
+		return _getFiles(modulesFolder, Paths.get(slashed, slot), filter);
+	}
+
+	private static Path[] _getFilesFrom(File layeredPath, FileFilter filter) {
+		ArrayList<Path> list = new ArrayList<>();
+		File[] children = layeredPath.listFiles();
+
+		for (File child : children) {
```
Comment (christopherbryanboyd) [inline cli/src/main/java/com/liferay/blade/server/PortalWildFlyBundleFactory.java]: A Java 8 `Stream` might be simpler here than a for-loop.

File: cli/src/test/java/com/liferay/blade/cli/command/ServerCommandsTest.java [hunk 17]
```diff
@@ -177,6 +192,41 @@ public void run() {
 		Assert.assertFalse("Expected server start process to be destroyed.", pidProcess.isAlive());
 	}
 
+	private void _testServerStatus(boolean expectedServerStatus, String timeout) throws Exception {
```
Comment (christopherbryanboyd) [inline cli/src/test/java/com/liferay/blade/cli/command/ServerCommandsTest.java]: Can we move this into a separate Pull Request, addressing https://issues.liferay.com/browse/BLADE-338 (add "server status" command)?

**Comment by gamerson** [comment]
hey @simonjhy   can you send updates based on @christopherbryanboyd  code review?  thanks!

File: cli/build.gradle [hunk 18]
```diff
@@ -94,6 +94,7 @@ dependencies {
 	compile group: "org.gradle", name: "gradle-base-services-groovy", version: "3.0"
 	compile group: "org.gradle", name: "gradle-core", version: "3.0"
 	compile group: "org.gradle", name: "gradle-tooling-api", version: "3.0"
+	compile group: "org.zeroturnaround", name: "zt-process-killer", version: "1.8"
```
[the remark this answers is not in the document]
Comment (simonjhy) [reply cli/build.gradle]: ok, change it to 1.9

File: cli/src/main/java/com/liferay/blade/cli/command/AbstractServerCommand.java [hunk 19]
```diff
@@ -0,0 +1,169 @@
...
+
+	@Override
+	public void execute() throws Exception {
+		bladeCLI = getBladeCLI();
+
+		BaseArgs baseArgs = bladeCLI.getBladeArgs();
+
+		File baseDir = new File(baseArgs.getBase());
+
+		File gradleWrapperFile = BladeUtil.getGradleWrapper(baseDir);
```
[the remark this answers is not in the document]
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/cli/command/AbstractServerCommand.java]: I will migrate related codes from BLADE-313 to AbstractServerCommand.

File: cli/src/main/java/com/liferay/blade/cli/command/ServerStatusCommand.java [hunk 20]
```diff
@@ -0,0 +1,138 @@
...
+import com.liferay.blade.server.PortalBundle;
+
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.net.URLConnection;
+
+/**
+ * @author Simon Jiang
+ */
+public class ServerStatusCommand extends AbstractServerCommand<ServerStatusArgs> {
```
[the remark this answers is not in the document]
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/cli/command/ServerStatusCommand.java]: When I did 316 we don't have 338 issue. I will move it to 338 after this pull will be merged.

File: cli/src/main/java/com/liferay/blade/cli/command/ServerStartArgs.java [hunk 21]
```diff
@@ -42,17 +54,37 @@ public void setBackground(boolean background) {
...
 	private boolean _debug;
 
+	@Parameter(description = "Set server debug port", names = {"-p", "--port"})
+	private String _port;
+
 	@Parameter(description = "Tail a running server", names = {"-t", "--tail"})
 	private boolean _tail;
 
+	@Parameter(description = "Set server type", names = {"-y", "--type"})
+	private String _type;
```
[the remark this answers is not in the document]
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/cli/command/ServerStartArgs.java]: yes, we can remove it now

File: cli/src/main/java/com/liferay/blade/cli/util/FileUtil.java [hunk 22]
```diff
@@ -103,6 +106,38 @@ public static void deleteDirIfExists(Path dirPath) throws IOException {
...
+	public static boolean exists(File file) {
+		if ((file != null) && file.exists()) {
+			return true;
+		}
+
+		return false;
+	}
+
+	public static boolean exists(Path path) {
+		if (path != null) {
```
[the remark this answers is not in the document]
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/cli/util/FileUtil.java]: nio exists method does not check path is null, maybe we need to keep it.

File: cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java (hunk 1 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java]: ok

File: cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java (hunk 2 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/cli/util/JavaUtil.java]: ok

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java (hunk 13 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: ok, I change it to use getProperty()

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java (hunk 12 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: ok

File: cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java (hunk 11 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/LayeredModulePathFactory.java]: ok

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java (hunk 10 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: ok

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java (hunk 9 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: ok

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java (hunk 8 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: ok

File: cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java (hunk 7 above)
Comment (simonjhy) [reply cli/src/main/java/com/liferay/blade/server/AbstractPortalBundleFactory.java]: ok

File: cli/src/test/java/com/liferay/blade/cli/command/ServerCommandsTest.java (hunk 17 above)
Comment (simonjhy) [reply cli/src/test/java/com/liferay/blade/cli/command/ServerCommandsTest.java]: ok

**Comment by simonjhy** [comment]
ok
